src/core/ext/filters/client_channel/http_proxy.cc

```diff
@@ -161,53 +165,40 @@ bool HttpProxyMapper::MapName(const char* server_uri,
               "unable to split host and port, not checking no_proxy list for "
               "host '%s'",
               server_uri);
-      gpr_free(no_proxy_str);
     } else {
-      size_t uri_len = server_host.size();
-      std::vector<std::string> no_proxy_hosts =
-          absl::StrSplit(no_proxy_str, NO_PROXY_SEPARATOR, absl::SkipEmpty());
+      std::vector<absl::string_view> no_proxy_hosts =
+          absl::StrSplit(no_proxy_str.get(), ',', absl::SkipEmpty());
       for (const auto& no_proxy_entry : no_proxy_hosts) {
-        size_t no_proxy_len = no_proxy_entry.size();
-        if (no_proxy_len <= uri_len &&
-            gpr_stricmp(no_proxy_entry.c_str(),
-                        &(server_host.c_str()[uri_len - no_proxy_len])) == 0) {
+        if (absl::EndsWithIgnoreCase(server_host, no_proxy_entry)) {
           gpr_log(GPR_INFO, "not using proxy for host in no_proxy list '%s'",
                   server_uri);
           use_proxy = false;
           break;
         }
       }
-      gpr_free(no_proxy_str);
-      if (!use_proxy) goto no_use_proxy;
+      if (!use_proxy) return no_use_proxy();
     }
   }
   server_target =
       MaybeAddDefaultPort(absl::StripPrefix(uri->path(), "/")).c_str();
-  grpc_arg args_to_add[2];
-  args_to_add[0] = grpc_channel_arg_string_create(
+  absl::InlinedVector<grpc_arg, 2> args_to_add;
+  args_to_add.push_back(grpc_channel_arg_string_create(
       const_cast<char*>(GRPC_ARG_HTTP_CONNECT_SERVER),
-      const_cast<char*>(server_target.c_str()));
+      const_cast<char*>(server_target.c_str())));
   if (user_cred != nullptr) {
     /* Use base64 encoding for user credentials as stated in RFC 7617 */
-    char* encoded_user_cred =
-        grpc_base64_encode(user_cred, strlen(user_cred), 0, 0);
+    auto encoded_user_cred =
+        UniquePtr<char>(grpc_base64_encode(user_cred, strlen(user_cred), 0, 0));
     std::string header =
```

Comment:
This needs to be declared outside of this scope, or else the entry in `args_to_add` will be pointing to already-freed memory when we call `grpc_channel_args_copy_and_add()`.

Reply:
That's exactly why the tests were failing. Done!